internal/uvm/cpugroups.go

```diff
@@ -0,0 +1,289 @@
+package uvm
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"github.com/Microsoft/hcsshim/internal/hcs"
+	"github.com/Microsoft/hcsshim/internal/log"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+)
+
+const (
+	CPUGroupNullID = "00000000-0000-0000-0000-000000000000"
+	MaxCPUGroupCap = 65536
+
+	// Default values in host for cpugroups
+	DefaultCPUGroupCap      = MaxCPUGroupCap
+	DefaultCPUGroupPriority = 1
+)
+
+var _HV_STATUS_INVALID_CPU_GROUP_STATE = errors.New("The hypervisor could not perform the operation because the CPU group is entering or in an invalid state.")
+
+// ReleaseCPUGroup unsets the cpugroup from the VM and attemps to delete it
+func (uvm *UtilityVM) ReleaseCPUGroup(ctx context.Context) error {
+	groupID := uvm.cpuGroupID
+	if err := uvm.unsetCPUGroup(ctx); err != nil {
+		return fmt.Errorf("failed to remove VM %s from cpugroup %s", uvm.ID(), groupID)
+	}
+
+	err := deleteCPUGroup(ctx, groupID)
+	if err != nil && err == _HV_STATUS_INVALID_CPU_GROUP_STATE {
+		log.G(ctx).WithField("error", err).Warn("cpugroup could not be deleted, other VMs may be in this group")
+		return nil
```

Comment:
Just to be sure, we mean to return nil here?

Reply:
Yes, I'll add a comment to explain better. But this error case happens if the cpugroup has other compute systems in it, in which case it can't be deleted.